:sparkles: Add progress bar and performance optimizations to analysis modes

tsanders-rh opened the pull request:
## Summary

Enhances kantra analysis with visual progress feedback and startup performance optimizations for both containerless and hybrid modes.

**Depends on:** #592 (Hybrid analysis mode)

## Changes

### Progress Bar Implementation
- **Containerless mode progress bar**: Real-time visual feedback showing percentage, completion bar, and current rule being processed
- **Hybrid mode progress bar**: Matches containerless UX with clean status messages and visual progress
- **Binary analysis support**: Progress tracking for JAR/WAR/EAR file analysis
- **`--no-progress` flag**: Disable progress bar for scripting/CI environments

### Performance Optimizations  
- **Parallel rule loading**: Load multiple rulesets concurrently instead of sequentially
- **Parallel startup tasks**: Run config validation, volume creation, and ruleset extraction concurrently
- Significantly reduces startup time when using multiple rulesets

### User Experience Improvements
- Clean initialization messages: "Running source/binary analysis...", "✓ Initialized providers", "✓ Loaded N rules"
- Cursor hiding/showing for cleaner progress rendering
- Suppressed container command output (volume/network creation)
- Fixed INFO message bleed that interfered with progress bar
- "Analysis complete!" message with results section

### Bug Fixes
- Fixed timeout and race condition in binary analysis
- Improved health check result handling

## Example Output

**Before:**
```
INFO[0000] Running source code analysis
... verbose log output ...
```

**After (with progress bar):**
```
Running source code analysis...
✓ Initialized providers: java, builtin
✓ Loaded 45 rules

Progress: [████████████████████████████████] 100% (45/45) | Processing: quarkus-springboot
Analysis complete!

Results: 127 incidents across 8 files
```

## Testing

- ✅ Tested with Java source code analysis
- ✅ Tested with binary (.war) analysis  
- ✅ Tested with multiple rulesets (parallel loading)
- ✅ Tested `--no-progress` flag for scripting
- ✅ Verified both containerless and hybrid modes

## Dependencies

- Requires analyzer-lsp v0.8.1-alpha for progress event support
- Depends on PR #592 for hybrid mode implementation

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Live progress UI for provider init, decompilation, rule loading/execution and report generation; opt-out via --no-progress.
  * Hybrid in-process analysis with improved host/container path handling and parallel startup.
  * Binary input auto-detection with tailored progress/messages.
  * Option to publish container ports and capture command output for richer diagnostics.

* **Bug Fixes**
  * Stronger provider health/startup checks, better error propagation and cleanup.
  * Wait-for-target/decompilation waits now support timeouts and clearer status reporting.

* **Docs**
  * Updated hybrid health-check/readiness guidance.

* **Chores**
  * Dependency bumps and logging improvements.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tsanders-rh commented on cmd/analyze.go:
```diff
@@ -144,166 +144,148 @@ func NewAnalyzeCmd(log logr.Logger) *cobra.Command {
+		} else {
+			operationalLog = logr.Discard()
+		}
+		err = analyzeCmd.GenerateStaticReport(ctx, operationalLog)
```
Outdated.

eemcmullan commented on cmd/analyze-hybrid.go:
```diff
@@ -28,9 +31,20 @@ import (
+// validateProviderConfig validates configuration before starting provider containers.
+// This catches configuration errors early and provides helpful error messages.
 func (a *analyzeCommand) validateProviderConfig() error {
+	// Validate Maven settings file if specified
```
Could this be moved to the analyze command preRun validate? This check is necessary in all modes. Maybe in the future it'd be nice to have a separate validate in the preRun that is only for Java provider config values.

tsanders-rh replied:
commit: 5eb4d21

eemcmullan commented on cmd/analyze-hybrid.go:
```diff
@@ -434,33 +505,90 @@ func (a *analyzeCommand) RunAnalysisHybridInProcess(ctx context.Context) error {
+		resultChan := make(chan startupResult, 3)
+
+		// Task 1: Validate provider configuration
+		wg.Add(1)
```
NIT: might be cleaner to have start-up tasks in a separate func

tsanders-rh commented on cmd/analyze-bin.go:
```diff
@@ -235,46 +413,55 @@ func (a *analyzeCommand) RunAnalysisContainerless(ctx context.Context) error {
+		reportPath := filepath.Join(a.output, "static-report", "index.html")
+		fmt.Fprintf(os.Stderr, "  Report: file://%s\n", reportPath)
+		analysisLogPath := filepath.Join(a.output, "analysis.log")
+		fmt.Fprintf(os.Stderr, "  Logs:   %s\n", analysisLogPath)
```
See commit: 550df65

  The output now shows:
  Results:
    Report: file:///path/to/output/static-report/index.html
    Analysis logs: /path/to/output/analysis.log

tsanders-rh commented on cmd/analyze-hybrid.go:
```diff
@@ -220,17 +244,13 @@ func (a *analyzeCommand) setupNetworkProvider(ctx context.Context, providerName
-		providerSpecificConfig["lspServerPath"] = JDTLSBinLocation
-		providerSpecificConfig["bundles"] = JavaBundlesLocation
+		providerSpecificConfig["lspServerPath"] = "/jdtls/bin/jdtls"
+		providerSpecificConfig["bundles"] = "/jdtls/java-analyzer-bundle/java-analyzer-bundle.core/target/java-analyzer-bundle.core-1.0.0-SNAPSHOT.jar"
```
See commit: 7969ab2

tsanders-rh commented on cmd/analyze-bin.go:
```diff
@@ -216,15 +304,105 @@ func (a *analyzeCommand) RunAnalysisContainerless(ctx context.Context) error {
+	operationalLog.Info("[TIMING] Starting rule execution")
+	operationalLog.Info("evaluating rules for violations. see analysis.log for more info")
+
+	// Create progress reporter (or noop if disabled)
```
Great suggestion! Extracted the progress reporter setup (~77 lines duplicated in both files) into a
  shared `setupProgressReporter()` function.

  See commit: f3f7121

tsanders-rh commented:
@eemcmullan - Thanks for review.  All updated per comments.  Waiting on CI, but ready for you to recheck.  Thanks!

shawn-hurley commented on cmd/analyze-bin.go:
```diff
@@ -97,8 +134,11 @@ func (a *analyzeCommand) RunAnalysisContainerless(ctx context.Context) error {
-	consoleHook := &ConsoleHook{Level: logrus.InfoLevel, Log: a.log}
-	logrusAnalyzerLog.AddHook(consoleHook)
+	// but only if progress is disabled (to avoid interfering with progress bar)
+	if a.noProgress {
```
Should this go to a file?

tsanders-rh replied:
318060f

shawn-hurley commented on cmd/analyze-hybrid.go:
```diff
@@ -576,36 +750,85 @@ func (a *analyzeCommand) RunAnalysisHybridInProcess(ctx context.Context) error {
+	var ruleWg sync.WaitGroup
+	resultChan := make(chan ruleLoadResult, len(a.rules))
+
+	// Load each ruleset in parallel
```
I think that this is something that we should consider adding the analyzer-lsp library, this could be helpful for all the integrators (addon/hub, kai, kantra). I don't think we should stop it from going here, but just wanted to call that out.

tsanders-rh commented:
Thanks for the feedback, @shawn-hurley! I've addressed both of your comments:

## 1. Encapsulating `noProgress` checks

I've created a `ProgressMode` type in `cmd/progress_mode.go` that encapsulates all the `noProgress` logic. This provides better abstraction and makes the code more maintainable.

**Key methods:**
- `OperationalLogger(log)` - Returns the logger or `logr.Discard()` based on progress mode
- `ShouldAddConsoleHook()` - Determines if console hooks should be added
- `HideCursor()` / `ShowCursor()` - Terminal cursor control
- `Printf()` / `Println()` - Conditional output to stderr
- `IsEnabled()` / `IsDisabled()` - Query progress state

**Before:**
```go
var operationalLog logr.Logger
if a.noProgress {
    operationalLog = a.log
} else {
    operationalLog = logr.Discard()
}

if !a.noProgress {
    fmt.Fprintf(os.Stderr, "\033[?25l")
    defer fmt.Fprintf(os.Stderr, "\033[?25h")
}
```

**After:**
```go
progressMode := NewProgressMode(a.noProgress)
operationalLog := progressMode.OperationalLogger(a.log)

progressMode.HideCursor()
defer progressMode.ShowCursor()
```

This refactoring affected:
- `cmd/analyze-bin.go`: 11 instances replaced
- `cmd/analyze-hybrid.go`: 11 instances replaced

## 2. Logging discarded output with verbosity control

I've added helper methods in `cmd/command-context.go` to conditionally capture and log container command output based on verbosity level:

- `setupCommandOutput(cmd)` - Configures stdout/stderr (discard or capture)
- `logCommandOutput(cmdName, stdout, stderr)` - Logs captured output

**Implementation:**
```go
func (c *AnalyzeCommandContext) setupCommandOutput(cmd *exec.Cmd) (stdout, stderr *bytes.Buffer) {
    if c.log.V(2).Enabled() {
        // Capture output for verbose logging
        stdout = &bytes.Buffer{}
        stderr = &bytes.Buffer{}
        cmd.Stdout = stdout
        cmd.Stderr = stderr
        return stdout, stderr
    }
    // Discard output to keep console clean
    cmd.Stdout = io.Discard
    cmd.Stderr = io.Discard
    return nil, nil
}
```

**Usage:**
```go
cmd := exec.Command(Settings.ContainerBinary, args...)
stdout, stderr := c.setupCommandOutput(cmd)
err := cmd.Run()
c.logCommandOutput("network create", stdout, stderr)
```

**Behavior:**
- **Normal mode:** Output discarded (clean console)
- **Verbose mode (`-v 2`):** Output captured and logged for debugging

Applied to:
- `createContainerNetwork()`
- `createContainerVolume()`

All tests pass and the changes maintain backward compatibility.

shawn-hurley reviewed (approved):
@eemcmullan Will leave for you to merge when you are good to go!
